### DISCUSSION TITLE
Extreme Roles専用大型アドオン「Extreme Modes」構想案

### DISCUSSION THREAD

**Discussion by yukieiji** [yukieiji/ExtremeRoles#223] [Ideas]
公式がモードを切り替えを作ってくれたので作れるなら作りたいよね(正直マップ作るより楽な気がする)

- モード案とかそれに欲しい"現在実装されている"役職とか募集しておきます

### 注意事項
- モードに対しての新規役職案はここでもチケットでも募集していません
- 一部コード的な話も出てきます
- この構想案は頓挫する可能性をはらんでいます(現段階だと可能であると判断しています)
  - 可能であると決定された場合このディスカッションはチケットになります

**Comment by yukieiji** [comment]
基本的な依存関係
- Extreme Roles：公式で提供されているモードに対しての役職追加と対応を行う
- Extreme Modes：Extreme Rolesに提供されているAPIを用いて専用の新規モード実装や専用モードに対応した役職を追加

基本的にExtreme ModesではAmongUs本体のモードには一切触れないし改造もしない、AmongUs本体のモードにふれるのはExtreme Roles

**Comment by melonjagaimo** [comment]
モード案「恋人たちと横恋慕」
使用するExRの役職：ラバーズ、ヤンデレ
恋愛が蔓延するゲームに、横恋慕中のヤンデレが乗り込んできた。
ヤンデレに恋人を奪われる前に恋人たちはヤンデレを追放できるのか？

・プレイヤーの人数が奇数の場合、ヤンデレ1人または3人を除いたすべてのプレイヤーが2人1組のラバーズになる。偶数の場合はヤンデレ2人を除いたすべてのプレイヤーが2人1組のラバーズになる。奇数の場合のヤンデレの人数はオプションで変更可能。

・インポスターの人数は0。

・このモードのラバーズはExRのラバーズのオプション設定にかかわらず、以下の仕様を持つ。
１：クルー陣営である
２：ラバーズは他の役職を持たない
３：相方が死ぬと自決し、ニュートラル陣営化（いわゆる未亡人状態）はしない
４：通常のクルーと同じ視界設定で、停電の効果を受ける

・このモードのヤンデレはExRのヤンデレのオプション設定に依存する。（仕様上可能なら、ヤンデレの想い人はラバーズの中から選ぶ。）

**Reply by yukieiji** [reply]
> ・インポスターの人数は0。

バニラの仕様上インポスターに人数を通常サーバー等で0人とすることは出来ないです

全体的にゲームシステムが変わるほどのモードとして機能してない気がします、それはルールとして作るべきなのではないでしょうか

**Reply by melonjagaimo** [reply]
分かりました

**Comment by melonjagaimo** [comment]
ゲームシステムが変わるって、具体的にどのようなことを考えていますか？
自分が思い浮かぶのは…
・特定の行動をとった場合にゲームが即座に終了する（例：シェリフのプレイヤーが死亡するとインポスター陣営勝利になる）

・通常の勝利条件とは異なる勝利条件の実装、および通常の勝利条件の無効化（例：インポスター陣営が全滅するとインポスター陣営の勝利になる。クルー陣営はインポスターにおびえながらタスクをしなければならない）

・引き分けの実装（ローカルルールの一つである探偵アモアスでは探偵役以外のクルー全員のタスク完遂で引き分けになる）

・ゲームオプションを無視して独自のオプション設定を適用する（例：ゲームオプション上では投票スキップは可能だが、特定のモードだと不可能になる）

あたりですね

**Reply by yukieiji** [reply]
> ゲームシステムが変わるって、具体的にどのようなことを考えていますか？

基本的にはクラシックモードで考慮されていない挙動をすること全般です
例えばHideNSeekモードでは死体は通報できない、会議ボタンを押すことが出来ない、クルーが全員エンジニア、エンジニアのベント使用に使用回数制限が存在する、インポスターとクルーでイントロの時間が違うなどクラシックモードで考慮されていない挙動を多数行っております
上記の「恋人たちと横恋慕」だとクラシックモードで考慮されていない挙動をすることがないという部分とそのモードに対して新しく役職を追加する事ができないと思われます。なのでモードで実装する必要性がないと感じています

> ・特定の行動をとった場合にゲームが即座に終了する（例：シェリフのプレイヤーが死亡するとインポスター陣営勝利になる）
・通常の勝利条件とは異なる勝利条件の実装、および通常の勝利条件の無効化（例：インポスター陣営が全滅するとインポスター陣営の勝利になる。クルー陣営はインポスターにおびえながらタスクをしなければならない）
・引き分けの実装（ローカルルールの一つである探偵アモアスでは探偵役以外のクルー全員のタスク完遂で引き分けになる）

ここらへんは全てクラシックモード(ExRのクラシック動作)では考慮されている挙動になります

> ・ゲームオプションを無視して独自のオプション設定を適用する（例：ゲームオプション上では投票スキップは可能だが、特定のモードだと不可能になる）

無視する意味が正直わからないです、オプションはモード毎に分かれているのに

**Reply by melonjagaimo** [reply]
分かりました

**Comment by UzomuZo** [comment]
モード案「宝探し」

できるかわかりませんが
・かくれんぼ同様に会議ボタン、ぷるぷる使用不可。
・インポスターのキル、ベントボタン使用不可。
　もしくはキル使用不可＋他プレイヤーエンジニア。
　あるいはキルクールを制限時間以上に設定。
・ドールの欠片のように宝がマップのランダム位置に出現。
・種類によってポイントが違う宝を集めてポイント加算。
・制限時間終了後にポイントが多い人が勝利。

**Reply by yukieiji** [reply]
できるとは思われますがクラシック等の他の役職を入れるメリット等があんまりないのが問題点といえば問題点でしょうか

**Reply by UzomuZo** [reply]
>できるとは思われますがクラシック等の他の役職を入れるメリット等があんまりないのが問題点といえば問題点でしょうか

たしかに他の役職は使えないですね。
ただキルできるようにすると個人戦では無限復活できないと最初にキルされた人がやることなくなるのでやっぱりチーム戦ですかね。

・かくれんぼ同様に会議ボタン、サボタージュ、ぷるぷる使用不可。
・ドールの欠片のように宝がマップのランダム位置に出現。
・種類によってポイントが違う宝を集めてポイント加算。

・クルー陣営
マップ上の宝を集める。
キルされると所持していた分のポイントが減るが、他のクルーが死体を調べると死体の残りのポイントが加算されて死体が消える。
制限時間まで一人でも生存、あるいは一定ポイントまで集まると勝利。

・インポスター陣営
マップ上の宝を集めることはできない。
キルすることで相手が所持しているポイントの一部を奪う(減らす)ことができる。(7割くらい？)
制限時間内にクルー陣営を全員キル。

・その他
制限時間と総獲得ポイント(もしくはゲージ)が画面に表示。

**Comment by Inumiya** [comment]
モード案「人外の宴」

基本的な仕様
- クルーのアサイン数１
- クルーはキルボタン、能力ボタンの対象外
- 誰がクルーでどの役職なのか、全てのプレイヤーから見える
- ゲームの終了条件は、通常モードと同じ

死なないクルー役職があれば通常モードでも実現可能ですが、死なないクルー役職を普通のルールに入れるとバランスを取るのが難しくなってしまうと思います。なので、特殊モード限定で死なないクルーを実装すると良いのかと思います。

モードとして実装するなら以下のように特殊なことが出来ても良いかと考えています
- 誰がクルーになるか設定できる
  - かくれんぼモードのインポスターのように指名できる
- インポスターを２陣営用意する
  - インポスター陣営Aとインポスター陣営Bは敵同士
  - アサインルールに関わるので難しそう
- 第４の陣営を用意する
  - 第３陣営と違って、インポスターのように仲間がいるのが前提
  - 新たに役職をいくつか実装する必要がある

**Reply by yukieiji** [reply]
> モード案「人外の宴」
基本的な仕様
クルーのアサイン数１
クルーはキルボタン、能力ボタンの対象外
誰がクルーでどの役職なのか、全てのプレイヤーから見える
ゲームの終了条件は、通常モードと同じ
死なないクルー役職があれば通常モードでも実現可能ですが、死なないクルー役職を普通のルールに入れるとバランスを取るのが難しくなってしまうと思います。なので、特殊モード限定で死なないクルーを実装すると良いのかと思います。

基本的なAmongUsのシステム(当たり判定等)は共通の上でモードを作る(かくれんぼとかでもそこらへんはいじられてない)ので恐らく、これ専用の役職を作った場合クラシックでその役職を入れてプレイするのとほとんど変わらない気がします(勝利判定も変わらないので)

> 誰がクルーになるか設定できる
かくれんぼモードのインポスターのように指名できる

ここらへんはできると思われます

> インポスターを２陣営用意する
インポスター陣営Aとインポスター陣営Bは敵同士
アサインルールに関わるので難しそう

基本的なAmongUsのシステム(当たり判定等)は共通の上でモードを作るのでこれは恐らく出来ないと思います

> 第４の陣営を用意する
第３陣営と違って、インポスターのように仲間がいるのが前提
新たに役職をいくつか実装する必要がある

第四陣営周りの実装が入ると恐らくクラシックにも入るような気がします(かくれんぼのオプションに全ての陣営のオプションが有るのと同じように)

**Reply by Inumiya** [reply]
死なないクルー役職を入れるというより、どのクルー役職も死ななくなる、というイメージでした

クルー陣営は役職に関わらず
- キルの対象にならない
- 能力の対象にならない
- （書き忘れましたが）投票の対象にならない
- 役職が全員から見える

となります

ただやはりクラシックルールとの差別化が十分ではないので、何かしら特別な処理がいるなと思い、
インポスター以外の仲間がいる人外陣営を用意しようかと思ったのですが、クラシックにも反映されてしまうのですね
となると他に何かクラシックにない要素がいりますね

**Comment by yukieiji** [comment]
[Twitterでいってたこの案](https://twitter.com/yukieiji/status/1639145473549438976)を忘れないうちにメモとして

- モード名：未定
  - インポスターは指定人数
  - ラウンド形式
  - 1ラウンド毎にインポスター全員とクルー(指定人数、重複なし)がマップのランダムな箇所(重複なし)に解き放たれる
  - 終了条件
     - インポスターかクルー陣営が何ラウンドかの先取
     - もしくは第3陣営の特殊勝利条件が達成される
  - クルーは指定数のタスクを全て終わらせる、インポスターは指定人数キルをすることでラウンドごとの各陣営の勝利が決定
  - クルーのベース役職はエンジニア、インポスターはベントとサボはなし
  - 勝利決定後会議発生、30秒のインターバル後次のラウンド開始
